TESTING.md

```diff
@@ -2,12 +2,124 @@
-- [Environment variables](#Environment-variables)
+- [Tests split by feature set](#tests-split-by-feature-set)
+- [Adding new tests](#adding-new-tests)
+- [Binary testing](#binary-testing)
+- [Custom terraform scripts](#custom-terraform-scripts)
+- [Environment variables](#environment-variables)
+
+## Meeting prerequisites: Building the test environment
+
+To run the tests, your vCD needs to have the following:
+
+* (1) An external network, with at least one, but preferably two allocatable IP addresses;
+* (2) An organization
+    * (3) An organization administrator user
+    * (4) A Virtual Data Center (VDC) with enough resources to build several vApps and VMs, and preferably 2 storage profiles
+        * (5) an Edge gateway with a default gateway
+    * (6) A catalog
+    * (7) A vApp template in the catalog
+
+Optionally, to run tests in [go-vcloud-director](https://github.com/vmware/go-vcloud-director), you will also need one or
+two Org VDC networks and a media item.
+
+
+This is how you can build the environment using Terraform itself.
+
+The above entities are defined in the test configuration file (see `sample_vcd_test_config.json` for the full list of
+entities to provide). Here is an excerpt of where these names come from:
+
+```
+{
+  "vcd": {
+     "org": "datacloud",                    // (2)
+     "vdc": "datacloudvdc",                 // (4)
+     "catalog": {
+       "name": "datacloudcat",              // (6)
+       "catalogItem": "photon-hw11"         // (7)
+     },
+     [...]
+   },
+   "networking": {
+     "edgeGateway": "gwdatacloud",           // (5)
+     "externalNetwork": "extnetdatacloud",   // (1)
+     [...]
+     }
+   },
+   [...]
+ }
+
+```
+
+The vCD environment builder uses the test system information to create these entities. When triggered, it will take the
+names of the entities and some of the configuration parameters from the regular test configuration file. It will then
+integrate such data with additional information that is only used for the environment build, and it's listed under the
+`testEnvBuild` section in the file:
+
+```
+{
+   [...]
+  "testEnvBuild": {
+    // Indispensable data, which is not normally used in the tests but is mandatory to create the
+    // environment
+
+    "gateway": "GATEWAY_IP",
+    "netmask": "255.255.224.0",
+    "externalNetworkStartIp": "STARTING_IP",
+    "externalNetworkEndIp": "END_IP",
+    "dns1": "DNS_IP",
+
+    // Optional data. If not provided, the corresponding values
+    // from the rest of the configuration file will be used
+
+    "dns2": "8.8.4.4",
+    "externalNetworkPortGroup": "VM Network",
+    "externalNetworkPortGroupType": "NETWORK",
+    "ovaPath": "/path/to/photon-hw11-3.0-26156e2.ova",
+    "media_path":  "/path/to/test.iso",
+
+    // extra data. If not provided, these resources will not be created
+    // These items can be used for go-vcloud-director testing
+
+    "mediaName": "test_media",
+    "storageProfile2": "Development",
```

Comment:
With two different storage profiles `mate test-env-build` created the environment, but `make testacc` got this failure:

```
--- FAIL: TestAccVcdVAppVmMulti (533.18s)
    testing.go:568: Step 0 error: errors during apply:

        Error: error changing memory size: &errors.errorString{s:"error changing memory size: API Error: 400: [ 276b01eb-6993-4056-9f20-482e19268cc6 ] Cannot use multiple storage profiles in a fast-provisioned VDC \"myvdc\" for VM \"TestAccVcdVAppVmVm1\"."}

          on /var/folders/0g/mmbktw7j0vl9d1hhrxbpg05r0000gq/T/tf-test346736688/main.tf line 32:
          (source code not available)
```

go-vcloud-director.log snippet (assuming this is the one):

```
2019/06/21 17:23:04 --------------------------------------------------------------------------------
2019/06/21 17:23:04 Request caller: vcd.resourceVcdVAppVmCreate-->vcd.resourceVcdVAppVmUpdateExecute-->govcd.(*VM).ChangeMemory
Size-->govcd.executeRequest
2019/06/21 17:23:04 PUT https://bos1-vcd-sp-static-201-150.eng.vmware.com/api/vApp/vm-8d8d382b-a01b-452a-80d3-b3bd3d2a87d3/virt
ualHardwareSection/memory
2019/06/21 17:23:04 --------------------------------------------------------------------------------
2019/06/21 17:23:04 Request data: [1107] <?xml version="1.0" encoding="UTF-8"?>
  <vcloud:Item xmlns:rasd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData" xmlns:vcloud="h
ttp://www.vmware.com/vcloud/v1.5" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" vcloud:href="https://bos1-vcd-sp-static
-201-150.eng.vmware.com/api/vApp/vm-8d8d382b-a01b-452a-80d3-b3bd3d2a87d3/virtualHardwareSection/memory" vcloud:type="applicatio
n/vnd.vmware.vcloud.rasdItem+xml">
      <rasd:AllocationUnits>byte * 2^20</rasd:AllocationUnits>
      <rasd:Description>Memory Size</rasd:Description>
      <rasd:ElementName>1024 MB of memory</rasd:ElementName>
      <rasd:InstanceID>5</rasd:InstanceID>
      <rasd:Reservation>0</rasd:Reservation>
      <rasd:ResourceType>4</rasd:ResourceType>
      <rasd:VirtualQuantity>1024</rasd:VirtualQuantity>
      <rasd:Weight>0</rasd:Weight>
      <vcloud:Link href="https://bos1-vcd-sp-static-201-150.eng.vmware.com/api/vApp/vm-8d8d382b-a01b-452a-80d3-b3bd3d2a87d3/virtualHardwareSection/memory" type="application/vnd.vmware.vcloud.rasdItem+xml" rel="edit"></vcloud:Link>
  </vcloud:Item>
2019/06/21 17:23:04 Req header:
2019/06/21 17:23:04     X-Vcloud-Authorization: [********]
2019/06/21 17:23:04     Accept: [application/*+xml;version=27.0]
2019/06/21 17:23:05 ################################################################################
2019/06/21 17:23:05 Response caller vcd.resourceVcdVAppVmUpdateExecute-->govcd.(*VM).ChangeMemorySize-->govcd.executeRequest-->govcd.executeRequestCustomErr-->govcd.checkRespWithErrType-->govcd.ParseErr-->govcd.decodeBody
2019/06/21 17:23:05 Response status 400 Bad Request
2019/06/21 17:23:05 ################################################################################
2019/06/21 17:23:05 Response header:
2019/06/21 17:23:05     Date: [Fri, 21 Jun 2019 14:23:05 GMT]
2019/06/21 17:23:05     X-Vmware-Vcloud-Request-Id: [276b01eb-6993-4056-9f20-482e19268cc6]
2019/06/21 17:23:05     X-Vmware-Vcloud-Ceip-Id: [0efa6401-d30a-4a45-892d-1a3c3441e44a]
2019/06/21 17:23:05     X-Vcloud-Authorization: [********]
2019/06/21 17:23:05     Content-Type: [application/vnd.vmware.vcloud.error+xml;version=27.0]
2019/06/21 17:23:05     X-Vmware-Vcloud-Request-Execution-Time: [229]
2019/06/21 17:23:05 Response text: [15189] <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<Error xmlns="http://www.vmware.com/vcloud/v1.5" xmlns:ovf="http://schemas.dmtf.org/ovf/envelope/1" xmlns:vssd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_VirtualSystemSettingData" xmlns:common="http://schemas.dmtf.org/wbem/wscim/1/common" xmlns:rasd="http://schemas.dmtf.org/wbem/wscim/1/cim-schema/2/CIM_ResourceAllocationSettingData" xmlns:vmw="http://www.vmware.com/schema/ovf" xmlns:ovfenv="http://schemas.dmtf.org/ovf/environment/1" xmlns:vmext="http://www.vmware.com/vcloud/extension/v1.5" xmlns:ns9="http://www.vmware.com/vcloud/versions" stackTrace="com.vmware.vcloud.api.presentation.service.VmStorageClassSpanningNotAllowedException: Cannot use multiple storage profiles in a fast-provisioned VDC &quot;myvdc&quot; for VM &quot;TestAccVcdVAppVmVm1&quot;.&#xA;&#x9;at com.vmware.vcloud.common.validate.ValidationUtil.validateStorageClassSpanningAllowed(ValidationUtil.java:664)&#xA;&#x9;at com.vmware.ssdc.backend.services.impl.VAppManagerImpl.createUpdateVmParams(VAppManagerImpl.java:3531)&#xA;&#x9;at ...
```

Reply:
This is a configuration issue. The storage profile needed for the test is the main one (usually `*`), while `Development` is only needed for tests related to managing storage profiles.

I tried with two different vCD, and I got a pass always.